File: problems_051to075/problem_061/solution.py
import os
import sys
from math import ceil, sqrt

PROJECT_PATH = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.append(PROJECT_PATH)

from measurement import Stopwatch
from toolbox import nth_polygonal_number


class Solution(Stopwatch):
    def __init__(self):
        super().__init__()

    def execute(self):
        answer = None

        polygonal_numbers = {}

        def find_cyclic_numbers(overlapping_numbers, polygonal_angles):
            last_number = overlapping_numbers[-1]

            for angle in polygonal_angles:
                for num in polygonal_numbers[angle]:
                    if str(last_number)[2:] == str(num)[:2]:
                        if len(polygonal_angles) == 1:
                            first_number = overlapping_numbers[0]

                            if str(num)[2:] == str(first_number)[:2]:
                                return [*overlapping_numbers, num]

                        else:
                            copy_of_angles = polygonal_angles[:]
                            copy_of_angles.remove(angle)
                            cyclic_numbers = find_cyclic_numbers([*overlapping_numbers, num], copy_of_angles)

                            if cyclic_numbers is not None:
                                return cyclic_numbers

            return None

        for angle in range(3, 9):
            n_limit = ceil(sqrt(20000/(angle-2) + 0.25) + 0.5)
            polygonal_numbers[angle] = [nth_polygonal_number(angle, n) for n in range(1, n_limit)]
            polygonal_numbers[angle] = [num for num in polygonal_numbers[angle] if num >= 10**3 and num < 10**4]

        for num in polygonal_numbers[8]:
            cyclic_numbers = find_cyclic_numbers([num], [3, 4, 5, 6, 7])

            if cyclic_numbers is not None:
                answer = sum(cyclic_numbers)
                break

        return answer

if __name__ == '__main__':
    soln = Solution()
    result = soln.execute()
    print(result)

File: problems_051to075/problem_061/solution.ts
import { SolutionFunction, Dictionary } from '@/lib/types';
import { range, sum, nthPolygonalNumber } from '@/lib/toolbox';

const solution: SolutionFunction = () => {
  let answer: ReturnType<SolutionFunction> = null;

  const polygonalNumbers: Dictionary<number[]> = {};

  const findCyclicNumbers = (overlappingNumbers: number[], polygonalAngles: number[]): number[] => {
    const [lastNumber] = overlappingNumbers.slice(-1);

    for (let angleIndex = 0; angleIndex < polygonalAngles.length; angleIndex += 1) {
      const angle = polygonalAngles[angleIndex];

      for (let index = 0; index < polygonalNumbers[angle].length; index += 1) {
        const num = polygonalNumbers[angle][index];

        if (String(lastNumber).slice(2) === String(num).slice(0, 2)) {
          if (polygonalAngles.length === 1) {
            const firstNumber = overlappingNumbers[0];

            if (String(num).slice(2) === String(firstNumber).slice(0, 2)) {
              return [...overlappingNumbers, num];
            }
          }
          else {
            const otherAngles = polygonalAngles.filter(
              (polygonalAngle) => (polygonalAngle !== angle),
            );
            const cyclicNumbers = findCyclicNumbers([...overlappingNumbers, num], otherAngles);

            if (cyclicNumbers.length) {
              return cyclicNumbers;
            }
          }
        }
      }
    }

    return [];
  };

  for (let angle = 3; angle <= 8; angle += 1) {
    const limit = Math.ceil(Math.sqrt(20000 / (angle - 2) + 0.25) + 0.5);

    polygonalNumbers[angle] = range(1, limit)
      .map((n) => nthPolygonalNumber(angle, n))
      .filter((num) => (num >= 10 ** 3 && num < 10 ** 4));
  }

  polygonalNumbers[8].every((num) => {
    const cyclicNumbers = findCyclicNumbers([num], [3, 4, 5, 6, 7]);

    if (cyclicNumbers.length) {
      answer = sum(cyclicNumbers);
    }

    return (cyclicNumbers.length === 0);
  });

  return answer;
};

export default solution;
